sdk/go/pulumi/types.go

```diff
@@ -64,6 +78,8 @@ type OutputState struct {
 	mutex sync.Mutex
 	cond  *sync.Cond
 
+	join *sync.WaitGroup // the wait group associated with this output, if any.
```

Comment:
OK so this is the meat. Since Output.getState() *OutputState, we now have `Output -> WaitGroup`, so presumably Apply chains propagate the wait group. 

I will use my beginner status to ask stupid questions freely.

# Zipping outputs across wait groups

What happens when zip join two outputs with different waitgroups?

```
All(a, b) // now what?

Given a.getState().join != b.getState().join
```

# Monadic bind across wait groups

Similar scenario, suppose you have `o Output<A>` and `f func(x A) Output<B>`, and we're computing `o2 = o.ApplyT(f)` to get `o2 Output<B>`.

```
func TestMonadicBind(t *testing.T) {
	o, resolveO, _ := NewOutput()
	p, resolveP, _ := NewOutput()

	// assume o.getState().join != p.getState().join

	resolveO(1)
	resolveP("OK")

	f := func(xval interface{}) Output {
		return p
	}

	o2 := o.ApplyT(f)

	// what is o2.getState().join ?

	v, known, secret, deps, err := await(o2)

	t.Errorf("v=%v known=%v secret=%v deps=%v err=%v", v, known, secret, deps, err)
}
```

If we have the answer to joining two wait groups we can apply the same answer here.


#  Which state to propagate

Is waitGroup the right (minimal) thing to propagate? Better than propagating Context?

Reply:
> Given a.getState().join != b.getState().join
> 
> Assume c = All(a, b) 
> 
> Now what do we have?
> 
> c.getState().join == ?

In this case we bump the counts on both a.join and b.join, create a new wait group for c.join, and fire off a goroutine that decrements the counts for a.join and b.join once c.join completes. This keeps a.join and b.join alive until all work associated with c (incl. any applies that are hung off of it) finishes. It's a little crude, but it gets the job done.

Reply:
> If we have the answer to joining two wait groups we can apply the same answer here.

Interesting, yeah. I don't think that these changes handle that case--I'll have to think this scenario over a bit more closely. I do think that we _can_ handle it, though, with some care.

Reply:
That said, I think that combining outputs from multiple contexts will be very uncommon, so I'm content to ship these changes without handling this case and follow up with a fix later.

Reply:
Your solution on `All` sounds clever. So it implies some method to combine wait groups `combineWG(a WG, b WG): WG` that produces c.join out of a.join and b.join. 

For the monad case, presumably you can use the same method, but the big caveat is lifecycle. `f(x).join` is not going to be there util the computation resolves, essentially until the Output resolves. 

If you had a magic method that did time travel:

```
createDeferredWaitgroup(wg: Output<WaitGroup>): WG
```

Then you could hook things up with:

```
combineWG(x.join, createDeferredWaitgroup(f(x).Apply(\y-> y.join)))
```

Reply:
I think that we could update the `fulfillValue` method to use a solution similar to `All`. Instead of deferring `o.join.Done()` [here](https://github.com/pulumi/pulumi/pull/6983/files#diff-192ea26149bdd1fec39b6476e5896ad8247830cfac2b0dbc47cf1f9f73313cb0R127-R130), we would fire off a grouting iff the fulfilled value is an `output` with a different wait group.

Reply:
Tried a few test variations involving sleeping in apply; current implementation is holding up.